### PR TITLE
Bug 1712562: pin the 4.1 GA candidate RHCOS build boot images

### PR DESCRIPTION
**Pull request by imcleod**
Putting WIP for now so we can discuss the URI (and destination branch) here

### CODE REVIEW

**Comment by wking** [comment]
I bumped the PR subject to reference https://bugzilla.redhat.com/show_bug.cgi?id=1712562

File: data/data/rhcos.json [hunk 1]
```diff
@@ -1,104 +1,111 @@
...
+            "hvm": "ami-0c1d2b5606111ac8c"
         },
         "us-west-2": {
-            "hvm": "ami-096cb19297a5c2726"
+            "hvm": "ami-00745fcbb14a863ed"
         }
     },
-    "baseURI": "https://releases-art-rhcos.svc.ci.openshift.org/art/storage/releases/ootpa/410.8.20190508.1/",
-    "buildid": "410.8.20190508.1",
+    "baseURI": "https://releases-art-jenkins.cloud.paas.upshift.redhat.com/storage/releases/ootpa/410.8.20190520.0/",
```
Comment (yuqi-zhang) [inline data/data/rhcos.json]: The original URL is the public redirected one. Although if this is the GA candidate, do we want a more "official" location? If not the original baseURI doesn't need to be changed

File: data/data/rhcos.json [hunk 2]
```diff
@@ -1,104 +1,111 @@
...
+    "buildid": "410.8.20190520.0",
     "images": {
+        "aws": {
```
Comment (yuqi-zhang) [inline data/data/rhcos.json]: We didn't ship this explicitly in the past, maybe we should continue to remove this section?

File: data/data/rhcos.json (hunk 1 above)
Comment (cgwalters) [inline data/data/rhcos.json]: Change the `baseURI` back please, see the previous PRs here.

**Comment by wking** [comment]
9f1e50dd8129 looks good to me.

/approve
/hold

Waiting on folks to green-light [the issue][1] for 4.1.0.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1712562

File: data/data/rhcos.json (hunk 1 above)
Comment (wking) [reply data/data/rhcos.json]: no more `baseURI` with 9f1e50d.

File: data/data/rhcos.json (hunk 2 above)
Comment (wking) [reply data/data/rhcos.json]: Gone with 9f1e50d.

File: data/data/rhcos.json (hunk 1 above)
Comment (wking) [reply data/data/rhcos.json]: No more `baseURI` with 9f1e50d.

**Comment by wking** [comment]
@tracyrankin set [the issue][1] target to 4.1.0 (why doesn't Bugzilla drop anchors so I can link these metadata changes).  So we're all set to go here once we get approval from @crawford or other lead that they're ok with how this PR is resolving the issue.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1712562

**Comment by crawford** [comment]
Why are we using anything other than the output of that script? How will the removal of those sections affect non-supported platforms? (gracefully error? panic?)

**Comment by wking** [comment]
> Why are we using anything other than the output of that script?

Because there's no way for the installer to access or otherwise expose the removed content, with the 
exception of libvirt, which we're not building anyway.  And removing the content avoids breaking in a CI-registry `baseURI` when folks should be [using the mirrors][1].

> How will the removal of those sections affect non-supported platforms?

You might be able to panic libvirt, but we aren't building for libvirt (I'll check).  We don't have consumers for the other removed data.

[1]: https://docs.openshift.com/container-platform/4.1/installing/installing_bare_metal/installing-bare-metal.html#%20installation-user-infra-machines-iso-installing-bare-metal

**Comment by wking** [comment]
Oh, and we can't use the mirrors for `baseURI`, because the 410.8.20190520.0 data hasn't been pushed up there yet.

**Comment by wking** [comment]
```console
$ git checkout origin/pr/1770
$ git describe --dirty
unreleased-master-995-g9f1e50dd8
$ TAGS=libvirt hack/build.sh
$ mkdir testing
$ cat <<EOF >testing/install-config.yaml
> apiVersion: v1
> baseDomain: installer.testing
> metadata:
>   name: wking
> platform:
>   libvirt:
>     URI: qemu+tcp://192.168.122.1/system
> pullSecret: |
>   {
>     ...REDACTED...
>   }
> EOF
$ ./bin/openshift-install --dir testing create cluster
INFO Fetching OS image: .                         
FATAL failed to fetch Terraform Variables: failed to generate asset "Terraform Variables": failed to get libvirt Terraform variables: failed to use cached libvirt image: Get : unsupported protocol scheme "" 
```
Not the greatest error message, but not a panic.

**Comment by wking** [comment]
[e2e-aws-upgrade][1]:

```
fail [k8s.io/kubernetes/test/e2e/upgrades/apps/job.go:63]: Expected
    <bool>: false
to be true
```

That's [a known issue][2].

/test e2e-aws-upgrade

[1]: https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_installer/1770/pull-ci-openshift-installer-release-4.1-e2e-aws-upgrade/12
[2]: https://bugzilla.redhat.com/show_bug.cgi?id=1708454

**Comment by crawford** [comment]
I'm having trouble understanding this statement:

> And removing the content avoids breaking in a CI-registry baseURI when folks should be using the mirrors.

What is the harm in leaving behind that URL? Is it just to prevent the accidental use of those CI artifacts?

**Comment by abhinavdahiya** [comment]
The images section cannot be removed:
- It breaks libvirt, for as much as we think that's developer only that's still important
- The images section is used for UPI vsphere, metal CI testing and those are 4.1 GA requirements.
    - https://github.com/openshift/release/blob/master/ci-operator/templates/openshift/installer/cluster-launch-installer-upi-e2e.yaml#L384
    - https://github.com/openshift/release/blob/master/ci-operator/templates/openshift/installer/cluster-launch-installer-metal-e2e.yaml#L262-L264
     And CI cannot use the upstream mirror locations.

/hold

**Comment by wking** [comment]
> What is the harm in leaving behind that URL? Is it just to prevent the accidental use of those CI artifacts?

Yeah.  If folks want to use these artifacts for official releases like what comes out of this branch, they should be pulling them from the mirrors, not the CI-cluster's forwarder.

> The images section cannot be removed:
>
> * It breaks libvirt, for as much as we think that's developer only that's still important

I don't really care either way for libvirt.  Folks running libvirt can supply their own QEMU URI regardless of what we do here.  I'd be happy with a built-in link to the mirrors for the libvirt folks, but since this isn't up on the mirrors yet that's not possible.  I'm ok with the CI-redirector link for the libvirt folks, because breaking them when the CI cluster happens to be down or the redirector is otherwise out of action won't break production users.

> * The images section is used for UPI vsphere, metal CI testing and those are 4.1 GA requirements.

Having these tests diverge from [our][1] [docs][2] is a potential source of CI/user divergence.  But good point, I'd forgotten about this "consumer API" :p.

> And CI cannot use the upstream mirror locations.

Why not?  Or is this just "we may make breaking changes to the bootimages that we need to preflight in CI", in which case I'm concerned about how we'd roll that out in the 4.1 docs.  Can we save breaking bootimage changes for 4.2?

I'm not strongly committed to this second commit.  If you're both ok with the first, we can drop it and get this landed without further discussion.  But if we do want to land a `baseURI` to support these other consumers, it seems like the CI-redirector is not where we want to be pointing.

[1]: https://docs.openshift.com/container-platform/4.1/installing/installing_vsphere/installing-vsphere.html#installation-vsphere-machines_installing-vsphere
[2]: https://docs.openshift.com/container-platform/4.1/installing/installing_vsphere/installing-vsphere.html#installation-vsphere-machines_installing-vsphere

**Comment by crawford** [comment]
I'm fine with just that first commit.

**Comment by wking** [comment]
> I'm fine with just that first commit.

Ok, I've dropped 9f1e50dd8 (talked to @imcleod about it earlier, and he's fine either way).

**Comment by abhinavdahiya** [comment]
> > What is the harm in leaving behind that URL? Is it just to prevent the accidental use of those CI artifacts?
> 
> Yeah. If folks want to use these artifacts for official releases like what comes out of this branch, they should be pulling them from the mirrors, not the CI-cluster's forwarder.
> 
> > The images section cannot be removed:
> > 
> > * It breaks libvirt, for as much as we think that's developer only that's still important
> 
> I don't really care either way for libvirt. 

Libvirt is an IPI style i do care about not breaking this way.

>Folks running libvirt can supply their own QEMU URI regardless of what we do here. I'd be happy with a built-in link to the mirrors for the libvirt folks, but since this isn't up on the mirrors yet that's not possible. I'm ok with the CI-redirector link for the libvirt folks, because breaking them when the CI cluster happens to be down or the redirector is otherwise out of action won't break production users.
> 

The goal is to treat IPI same, so any divergence we do there, i'm against that. LIbvirt needs to do the same as AWS, ie use the embedded.

> > * The images section is used for UPI vsphere, metal CI testing and those are 4.1 GA requirements.
> 
> Having these tests diverge from [our](https://docs.openshift.com/container-platform/4.1/installing/installing_vsphere/installing-vsphere.html#installation-vsphere-machines_installing-vsphere) [docs](https://docs.openshift.com/container-platform/4.1/installing/installing_vsphere/installing-vsphere.html#installation-vsphere-machines_installing-vsphere) is a potential source of CI/user divergence. But good point, I'd forgotten about this "consumer API" :p.

CI doesn't test mirrors, it tests the source code at that point of time. we promote the RHCOS pipeline to mirrros not the other way around...
And CI is not diverging from docs. Docs mention to use the "approved" bootimages, for master the embedded rhcos.json is the "approved".

> > And CI cannot use the upstream mirror locations.
> 
> Why not? Or is this just "we may make breaking changes to the bootimages that we need to preflight in CI", in which case I'm concerned about how we'd roll that out in the 4.1 docs. Can we save breaking bootimage changes for 4.2?
> 
> I'm not strongly committed to this second commit. If you're both ok with the first, we can drop it and get this landed without further discussion. But if we do want to land a `baseURI` to support these other consumers, it seems like the CI-redirector is not where we want to be pointing.


https://github.com/openshift/installer/pull/1770/commits/c270bf8c51793f3dc3153ac2937fba752608c762 lgtm

/hold cancel
/lgtm

**Comment by wking** [comment]
> Libvirt needs to do the same as AWS, ie use the embedded.

This is fine with me, I just think the CI redirector is the wrong choice for embedding.

> CI doesn't test mirrors, it tests the source code at that point of time. we promote the RHCOS pipeline to mirrros not the other way around.

We want to cover both the RHCOS pipeline and the mirror-facing docs.  Ideally we'd default to the mirrors for general release-branch CI, because that exercises the flow we're documenting for users.  If there's some issue with the documented flow (mirrors down, mirrors flaky, accidentally uploaded wrong content to the mirrors), mirror-facing CI would turn it up.  We also want to cover new RHCOS, but only for the installer jobs that are explicitly about gating RHCOS bootimage promotion.  I'm fine passing in overrides for the installer jobs in those cases, but I think the rest of the org should be using the mirrors for their `release-4.1` preflights.

And really, I'm less concerned about CI than I am about production consumers.  We can clobber whatever we need to clobber in CI.  But when we decide to give users a way to get bootimages that isn't "grep the source" or "read the docs" (#1399), I think we'll need to make the production/mirror vs. CI distinction more clear.  Or maybe we'll have moved beyond needing bootimages by then ;).

[1]: https://github.com/openshift/release/blob/6ca4a9371305b26407c90c2fe82a98c83b93591a/ci-operator/jobs/openshift/release/openshift-release-release-4.1-periodics.yaml#L6

**Pull request merged**